### DISCUSSION TITLE
Is there a way to speed up indexing?

### DISCUSSION THREAD

**Discussion by haok1402** [xhluca/bm25s#48] [Ideas & Requests]
Hi there!

Thanks for the great library. It's a user-friendly experience to perform retrieval in Python. May I ask is it possible to support indexing with multiple threads? As you can see from the screenshot, right now only one thread got used. Would really love a way to speed things up!

Thanks

![image](https://github.com/user-attachments/assets/5499add9-bc06-42d3-a137-a509a8ff0005)

**Comment by xhluca** [comment]
Thank you! Currently the library only supports one thread indexing, but it would be great to add multi-thread indexing in the future! I am not currently sure what's the best way to approach this, I'd probably need to spend some time to think about it.

That said, it seems quite slow that you can only process 400it/s. Even on Kaggle, it processes 17K/s: https://www.kaggle.com/code/xhlulu/benchmark-bm25s-msmarco

Can you perhaps try to run your code on kaggle to see if there's a bottleneck with your processing code, or with the server itself?

**Reply by haok1402** [reply]
Perhaps we could divide into multiple partitions, do indexing on its own in parallel, and combine the retrieval results based on the scoring? :eyes: I guess the problem is more inherent to the particular type of indexing required on my end. Each document would take roughly 512 unique tokens, the vocabulary size is about 373K, and there are 8M documents.

**Reply by xhluca** [reply]
`8,841,823` isn't that msmarco? Given 340M tokens in total (using the `bm25s.tokenize`), it should be 34 tokens per doc on avg. So not sure if you are encountering a different issue in that case. However, I still recommend trying out your scenario with kaggle first to see if you are observing something different from my notebook; after all, parallelizing 4x 400it/s could at most get you 1600 it/s, which is far from the 18K/s processed on kaggle (very old cpu, run on single core).

Regarding the suggestion for parallelization, the problem is that the document frequency needs to be computed first before computing the rest, otherwise a certain level of parallelization could work, but are still reliant on the computing doc frequency first. Also, i have found that the native python multiprocessing seems to struggle with multithreading, achieving only ~2x improvement when going from 1 to 4 cores, so it might not be that useful.
